Fixes #10478 - Add deb_repo type to Katello content

mdellweg opened the pull request:
This PR is part of the work to integrate deb-repositories into Katello.
It depends on the corresponding branches of pulp_deb and runcible:
https://github.com/ATIX-AG/pulp_deb/tree/feature/sync_release_structure
https://github.com/ATIX-AG/runcible/tree/feature/add_deb_type

jlsherrill commented:
Now that a pulp 2.14 beta has been released I will start testing and reviewing this!

sean797 commented on db/migrate/20170720142145_add_deb_release_parameters_to_repositories.rb:
```diff
@@ -0,0 +1,7 @@
+  def change
+    add_column :katello_repositories, :deb_releases, :string, limit: 255
+    add_column :katello_repositories, :deb_components, :string, limit: 255
+    add_column :katello_repositories, :deb_architectures, :string, limit: 255
```
Why create these, cant it use the columns already in this table? e.g `distribution_arch`, `distribution_version` ect..

mdellweg replied:
Debian (apt) repositories have different structure than yum.
They can contain packages for severals architectures, be divided into multiple components (think: main, contrib, non-free) and incorporate more than one Release.
That is why i introduced these new fields as plural terms.

mdellweg commented:
Dependency https://github.com/ATIX-AG/pulp_deb/tree/feature/sync_release_structure is merged in pulp_deb upstream.

Arceopteryx commented:
This feature holds high value to me. Is there any way I can help?
I have no Ruby experience however.

jlsherrill commented:
To clarify the status of this, we've mostly been waiting for a pulp pr (https://github.com/pulp/pulp_deb/pull/24) to be merged.  After a few months that has now been merged.  Now we can more easily review this PR, however we still can't merge it until some pulp release is available with the patch.    I'm not sure if this will be pulp 2.15 or if its possible this may come out sooner.

In addition this PR has been getting a lot of activity while new functionality is added.  This is great!  But at some point we need to slow down updates so that we can review what is there and stabilize the pr.  I'm planning on diving into to review this pretty heavily in the next few days.

jlsherrill commented:
If anyone does want to help, it should be possible to test with this pulp repository:  https://repos.fedorapeople.org/repos/pulp/pulp/testing/automation/2.15/stage/7Server/x86_64/    in conjunction with the katello dev environment

mdellweg commented:
From my point of view, this part of Katello-integration is roughly finished. If you like, we can call this a defined state, clean up what is already here, and defer further new stuff to a different branch based on this one.
BTW: Do you prefer commits to be fixed in place, or shall we collect fixup-commits and do the big interactive rebase in the end (assuming there will be a need for fixups...)? Or do you plan on one squashed commit anyway?

mdellweg commented:
@Arceopteryx thanks for your support. I is however kind of hard to break down tasks, if you only state what you can not do. Right now testing would be nice (theforeman/forklift may be a good starting point). Together with reporting issues (not the kind 'feature x is not yet implemented' of course.) Other than that, only sending pastry springs to my mind. ;)

tennityj commented:
Willing to test and give feedback. Spinning up dev environment with forklift. :D

Arceopteryx commented:
@mdellweg Thanks for the reponse! That is exactly what I will do then!

jlsherrill commented:
@mdellweg great to hear!   Sorry for the delays, I expect to get to reviewing this this week!  I'm fine with leaving the commits split out for now, but lets squash them as we get closer to merging.

jlsherrill commented on app/models/katello/deb.rb:
```diff
@@ -0,0 +1,89 @@
+    has_many :repositories, :through => :repository_debs, :class_name => "Katello::Repository"
+    has_many :repository_debs, :class_name => "Katello::RepositoryDeb", :dependent => :destroy, :inverse_of => :deb
+
+    has_many :content_facets, :through => :content_facet_applicable_debs, :class_name => "Katello::Host::ContentFacet"
```
I don't think this applicable_debs piece is useful for debs right now.  This requires special support in pulp to identify that some deb can be upgraded on a particular host, by knowing a) what is installed, b) what repositories a system is using ,and c) what deps are in that repo.  This is only supported in pulp for rpms and errata AFAIK.  Does the deb plugin add this for debs?   If it does then i'd expect to see this data being imported elsewhere in the PR, but i don't think i see that.

mdellweg replied:
Sounds reasonable. I never really understood the content-facet thing. So this one is owed to copying from RPM. Is there a good explanation lying around somewhere?

jlsherrill replied:
Some pulp documentation is here:  https://docs.pulpproject.org/dev-guide/integration/rest-api/consumer/applicability.html

but basically it lets you see what is available to a system to be upgraded.  I think it would be a great to have in the future for deb, but i'm assuming its not currently available in pulp_deb

jlsherrill commented on app/models/katello/deb.rb:
```diff
@@ -0,0 +1,89 @@
+    scoped_search :on => :arch, :complete_value => true
+    scoped_search :on => :epoch, :complete_value => true
+    scoped_search :on => :filename, :complete_value => true
+    scoped_search :on => :sourcerpm, :complete_value => true
```
doubt sourcerpm is useful here :)

jlsherrill commented on app/models/katello/deb.rb:
```diff
@@ -0,0 +1,89 @@
+    scoped_search :on => :name, :complete_value => true
+    scoped_search :on => :version, :complete_value => true
+    scoped_search :on => :release, :complete_value => true
+    scoped_search :on => :arch, :complete_value => true
```
looks like this should be 'architecture'

jlsherrill commented on app/models/katello/deb.rb:
```diff
@@ -0,0 +1,89 @@
+
+    scoped_search :on => :name, :complete_value => true
+    scoped_search :on => :version, :complete_value => true
+    scoped_search :on => :release, :complete_value => true
```
doesn't look like 'release' is  in the db migration for the deb table

jlsherrill commented on app/models/katello/deb.rb:
```diff
@@ -0,0 +1,89 @@
+    scoped_search :on => :version, :complete_value => true
+    scoped_search :on => :release, :complete_value => true
+    scoped_search :on => :arch, :complete_value => true
+    scoped_search :on => :epoch, :complete_value => true
```
doeesn't look like epoch is defined in the migration for the deb table

mdellweg replied:
Again, i learned something: The `scoped_search` automatically blends into the search field of the debs-Page. 🎒

jlsherrill commented on db/migrate/20170714144125_add_deb.rb:
```diff
@@ -0,0 +1,34 @@
+      t.string 'filename', :limit => 255
+      t.string 'checksum', :limit => 255
+      t.string 'version_sortable', :limit => 255
+      t.string 'description', :limit => 255
```
Reading https://www.debian.org/doc/debian-policy/#version  it looks like there could be value in a 'release' field, does pulp provide that?

mdellweg replied:
Do you mean 'revision'? Then we should also have epoch implemented. Right now, i think the version is imported as a whole into the `version_sorted` field. Is there a benefit in parsing that into three variables?

jlsherrill replied:
trying to think of a good argument against this way, but i can't really, so i guess its fine for now :) 👍

jlsherrill commented:
Here's what i see when syncing:

![deb1](https://user-images.githubusercontent.com/395077/32468175-2f47fd28-c31b-11e7-8d2c-f9c897384334.png)

I think the presenter ( app/lib/actions/pulp/repository/presenters/deb_presenter.rb) needs a bit of work.  I'm not sure how much you modified it from the yum presenter, but i'm guesing it nees some more customization.  The the yum presenter is pretty ugly and confusing code, so if you need any help, let me know!   Continuing to review :)

mdellweg commented:
rebased for conflicts.
@jlsherrill I improved the presenter a little. Do you know, what methods are supposed to be there?

jlsherrill commented:
[test]

jlsherrill commented:
@mdellweg arg sorry for the delay.  Going to keep a better eye on this now that the pulp 2.15 beta is out.   
Since these presenters include a helper "Helpers::Presenter::Base", they only need to define the humanized_details and progress methods.  Everything else is just useful methods for pulling apart the pulp task (although I will admit the yum presenter is quite a bit more complicated than it should be).

jlsherrill commented:
Seeing an error trying to remove a deb from a repo, the task fails, its pretty simple to fix though, this should do it:

```
diff --git a/app/lib/actions/katello/repository/remove_content.rb b/app/lib/actions/katello/repository/remove_content.rb
index c76d113..99baf5c 100644
--- a/app/lib/actions/katello/repository/remove_content.rb
+++ b/app/lib/actions/katello/repository/remove_content.rb
@@ -26,6 +26,8 @@ module Actions
                           Pulp::Repository::RemoveDockerManifest
                         when ::Katello::Repository::FILE_TYPE
                           Pulp::Repository::RemoveFile
+                        when ::Katello::Repository::DEB_TYPE
+                          Pulp::Repository::RemoveDeb
                         end
 
           uuids = content_units.map(&:uuid)
diff --git a/app/lib/actions/pulp/repository/remove_deb.rb b/app/lib/actions/pulp/repository/remove_deb.rb
new file mode 100644
index 0000000..0953fe6
--- /dev/null
+++ b/app/lib/actions/pulp/repository/remove_deb.rb
@@ -0,0 +1,15 @@
+module Actions
+  module Pulp
+    module Repository
+      class RemoveDeb < Pulp::Repository::AbstractRemoveContent
+        def content_extension
+          pulp_extensions.deb
+        end
+
+        def criteria
+          super.merge(fields: { :unit => ::Katello::Pulp::Deb::PULP_SELECT_FIELDS})
+        end
+      end
+    end
+  end
+end
diff --git a/app/models/katello/repository.rb b/app/models/katello/repository.rb
index a5aa1b8..31fd89a 100644
--- a/app/models/katello/repository.rb
+++ b/app/models/katello/repository.rb
@@ -639,6 +639,8 @@ module Katello
         self.ostree_branches -= units
       elsif file?
         self.files -= units
+      elsif deb?
+        self.debs -= units
       elsif docker?
         remove_docker_content(units)
       end
@@ -682,6 +684,8 @@ module Katello
         self.ostree_branches
       elsif file?
         self.files
+      elsif def?
+        self.debs
       else
         fail "Content type not supported for removal"
       end
```

jlsherrill commented on engines/bastion_katello/app/assets/javascripts/bastion_katello/content-views/details/views/content-view-details.html:
```diff
@@ -80,6 +80,13 @@ <h2 translate>{{ contentView.name }}</h2>
+      <li ng-class="{active: stateIncludes('content-view.repositories.deb')}"
+          ng-hide="contentView.composite">
+        <a ui-sref="content-view.repositories.deb.list" translate>
+          deb Repositories
```
I'd prefer this to be consistent with capitalization of the other tabs.  "Deb Repositories"

mdellweg replied:
Actually, i am not quite sure where to call them "Apt Repositories".

jlsherrill replied:
Thats a good point, anytime we refer to the repositories we should likely refer to them as "Apt Repositories", including when creating a new repo.

jlsherrill commented:
Few issues i've found:
* Content > Deb Packages > click a deb package > Library Repositories,  this seems to list all repos regardless of whether the package is in it or not
* Content > Deb Packages > click a deb package,  Path and checksum are empty.  I'm guessing these should come from pulp and be populated in the 'show.json.rabl', but it looks like thats not done
* Content > Deb Packages > click a deb package, The title of this is just the 'name' of the package, but i'd expect  'name-version' at a minimum, probably 'name-version_arch'  which seems like the standard in the filename.
* The sortable package version somewhat derived from the rpm packaging standards, and it appears debian has different standards.  For example:

1.5.0-1 is greater than 1.5.0~beta1, but with the included sortable string algorithm we're using, it thinks 1.5.0-1 is less than 1.5.0~beta1  

we could try to create a 2nd algorithm to convert deb versions into a sortable string, but i'm  fine with removing it for now (the column and the default sort on it).

Its getting very close!!!!

jlsherrill commented on app/models/katello/deb.rb:
```diff
@@ -0,0 +1,82 @@
+
+    def self.default_sort
+      order(:name).order(:epoch).order(:version_sortable).order(:release_sortable)
+    end
```
This isn't being used i don't think?  (it errors when i try to use it in the rails console, due to epoch)

jlsherrill commented on app/models/katello/deb.rb:
```diff
@@ -0,0 +1,82 @@
+      self.joins(:content_facets).
+        where("#{Katello::Host::ContentFacet.table_name}.host_id" => hosts).uniq
+    end
+  end
```
based on our discussions, lines 51 -81 aren't needed, correct?

mdellweg replied:
Definitely a copy from rpm relict.

mdellweg commented:
@jlsherrill reworked, rebased
I hope, i addressed all of the points you figured out. However, the Deb/Apt naming thing might still be inconsistent. If you spot some more places, let me know.
btw: Thanks for the delete patch!

jlsherrill commented:
[test default]
[test rails-5]

jlsherrill commented:
[test default]
[test rails-5]

jlsherrill commented:
@mdellweg want to go ahead and squash all the commits?

jlsherrill commented on app/models/katello/repository.rb:
```diff
@@ -739,6 +746,8 @@ def removable_unit_association
         self.ostree_branches
       elsif file?
         self.files
+      elsif def?
```
I think you mean deb? :)

mdellweg replied:
Yes i guess so. I just applied your diff as it looked sane... ;)

jlsherrill commented:
Seeing one small issue with uploading a debian package through the UI, it uploads fine but does not show up as having been uploaded (the count stays the same).  A small patch fixes it:

```
diff --git a/app/lib/actions/katello/repository/filtered_index_content.rb b/app/lib/actions/katello/repository/filtered_index_content.rb
index ddb12ac9e0..b5d74b998f 100644
--- a/app/lib/actions/katello/repository/filtered_index_content.rb
+++ b/app/lib/actions/katello/repository/filtered_index_content.rb
@@ -19,6 +19,8 @@ module Actions
             ::Katello::DockerManifest.import_for_repository(repo)
           elsif repo.file?
             ::Katello::FileUnit.import_for_repository(repo)
+          elsif repo.deb?
+            ::Katello::Deb.import_all(unit_ids)
           else
             ::Katello::Rpm.import_all(unit_ids)
           end
```

jlsherrill commented on app/models/katello/glue/pulp/repo.rb:
```diff
@@ -226,6 +230,16 @@ def puppet_importer_values(capsule)
+          components: self.deb_components,
+          architectures: self.deb_architectures
+        }
+        config.merge(importer_ssl_options(capsule))
```
this method got renamed in master to importer_connection_options

jlsherrill commented on app/lib/actions/katello/repository/clone_to_version.rb:
```diff
@@ -25,6 +25,9 @@ def plan(repositories, content_view_version, incremental = false)
                 end
+              elsif new_repository.deb?
+                plan_action(Repository::CloneDebContent, repository, new_repository, filters, !incremental,
+                            :generate_metadata => !incremental, :index_content => !incremental, :simple_clone => incremental)
```
I actually broke this!  notice that the repository variable isn't defined, I fixed it here:  https://github.com/Katello/katello/pull/7121

mdellweg replied:
I guess, we need to wait on this one then?

jlsherrill commented:
Overall, looking good.  Once these last issues are resolved, and everything is squashed, i think we can merge!

mdellweg commented:
@jlsherrill Squashed!
I put the last fix on a separate commit for now, because i think there is no way around rebasing, once  #7121 is merged.

jlsherrill commented:
@mdellweg it looks like there are ~4 tests failing: http://ci.theforeman.org/job/katello-pr-test/179/

jlsherrill commented:
and ignore the 'default' failure, rubocop is failing due to a new version.  @parthaa is fixing it here: https://github.com/Katello/katello/pull/7122

jlsherrill commented on db/migrate/20170714144125_add_deb.rb:
```diff
@@ -0,0 +1,34 @@
+class AddDeb < ActiveRecord::Migration
```
Rails 5.1 support is about to be merged in foreman, and requres some new syntax for migrations, see these examples:  https://github.com/Katello/katello/pull/7094/files  

Can you go ahead and add that here?

jlsherrill commented:
[test default]

jlsherrill commented:
There are two things I think that need to be done next:
1) update puppet-katello to install the debian pulp plugin
2) update the client docs https://theforeman.org/plugins/katello/nightly/installation/clients.html   to cover installing sub on a debian client

mdellweg commented:
@jlsherrill thank you for helping me along the way.

BloodyIron commented:
Can we also have this mean that Katello can have deb packages so we can install Katello on Ubuntu or Debian hosts that run Foreman? Managing Debian/Ubuntu hosts from Foreman is one thing, running Foreman on a Debian/Ubuntu host is another. There are those that want to keep an all-debian or all-ubuntu ecosystem, where possible. Currently Katello seems to only be available to install (as a content provider) on CentOS/RHEL hosts.
